source/darwin/nw_socket.c

```diff
@@ -893,101 +899,109 @@ static void s_process_connection_state_changed_task(struct aws_task *task, void
     nw_connection_t nw_connection = connection_args->nw_connection;
     nw_connection_state_t state = connection_args->state;
 
-    AWS_LOGF_INFO(
-        AWS_LS_IO_SOCKET,
-        "id=%p handle=%p: Apple network framework socket connection state changed to %d, nw error code : %d",
-        (void *)nw_socket,
-        (void *)nw_socket->os_handle.nw_connection,
-        connection_args->state,
-        connection_args->error);
+    /* Ideally we should not have a canceled task here, as nw_socket keeps a reference to event loop, therefore the
+     * event loop should never be destroyed before the nw_socket get destroyed. If we manually cancel the task, we
+     * should make sure we carefully handled the state change eventually, as the socket relies on this task to release
+     * and cleanup.
+     */
+    if (status != AWS_TASK_STATUS_CANCELED) {
 
-    switch (state) {
-        case nw_connection_state_cancelled: {
-            s_lock_base_socket(nw_socket);
-            struct aws_socket *socket = nw_socket->base_socket_synced_data.base_socket;
-            s_unlock_base_socket(nw_socket);
+        AWS_LOGF_INFO(
+            AWS_LS_IO_SOCKET,
+            "id=%p handle=%p: Apple network framework socket connection state changed to %d, nw error code : %d",
+            (void *)nw_socket,
+            (void *)nw_socket->os_handle.nw_connection,
+            connection_args->state,
+            connection_args->error);
 
-            s_lock_socket_synced_data(nw_socket);
-            s_set_socket_state(nw_socket, socket, CLOSED);
-            s_unlock_socket_synced_data(nw_socket);
+        switch (state) {
+            case nw_connection_state_cancelled: {
+                s_lock_base_socket(nw_socket);
+                struct aws_socket *socket = nw_socket->base_socket_synced_data.base_socket;
+                s_unlock_base_socket(nw_socket);
 
-            s_socket_release_internal_ref(nw_socket);
-        } break;
-        case nw_connection_state_ready: {
-            nw_path_t path = nw_connection_copy_current_path(nw_connection);
-            nw_endpoint_t local_endpoint = nw_path_copy_effective_local_endpoint(path);
-            nw_release(path);
-            const char *hostname = nw_endpoint_get_hostname(local_endpoint);
-            uint16_t port = nw_endpoint_get_port(local_endpoint);
+                s_lock_socket_synced_data(nw_socket);
```

Comment:
Per discussion, the socket can go away in between pulling it out here and the call to s_set_socket_state.  Also, per discussion, let's try and remove the set to the aws_socket.state completely inside s_set_socket_state